### ISSUE TITLE
In-text citation range processing fails under certain (very reproducible) circumstances.

### ISSUE THREAD

**Issue by dsifford** [Juris-M/citeproc-js#13]
I traced this issue and can confirm it is an issue with citeproc-js proper, rather than my implementation. 

Specifically, it's a failure within `CSL.Output.Queue.prototype.renderBlobs`

---
> **Note:** This is observed with AMA format, but likely can also be observed in others that use the same style of in-text citations.

### Expected behavior
If I have a single document with a range of citations from `1-3` and I re-cite the reference at position 3 before the range of citations, citation 3 should be shifted to location 1, but the range should not be affected.

**Example:**
This....
```
New citation here.

Old range here. 1-3
```

Should be this after inserting citation # 3 at the end of line 1...

```
New citation here. 1

Old range here. 1-3
```

### Actual behavior

![gif](http://i.giphy.com/l0MYQWV73IBIzF3OM.gif)

In other words, this is actually what happens....

Before
```
New citation here.

Old range here. 1-3
```
After
```
New citation here. 1

Old range here. 2,3,1
```

Hope you're doing well, Frank. 😄

**Comment by dsifford** [comment]
Adding more notes as I continue tracing backwords...

It looks like `myBlobs` of `CSL.getCitationCluster` is not ordered correctly.  This is index 0 of `myBlobs`...

![image](https://cloud.githubusercontent.com/assets/5240018/17792750/af5aef02-6570-11e6-8348-504bca9cf384.png)



Going to continue tracing backwards... standby

**Comment by dsifford** [comment]
Path stop at `this.output = new CSL.Output.Queue(this);` inside `CSL.getCitationCluster`...

Not much documentation on what's going on in `src/` for this, so I'm unfortunately not able to offer much help here.

**Comment by dsifford** [comment]
Update: It's not just me, I can reproduce this in your example as well..

![gif](http://i.giphy.com/3o7TKxwTCfd4a4x4TC.gif)

**Comment by fbennett** [comment]
Thanks for reporting this critter. It's a good one.

I've found the bug. It was correctly identifying citations affected by
a leading reinsert, but it was fixing only disambiguation issues, not
resorting the citation. I'm pretty sure this affected any style that
requires citation sorting (not just numeric ones). Very glad to get it
squashed!

Should be fixed now, in release tag 1.1.114.

**Comment by fbennett** [comment]
I've been watching tracker activity in ABT. Really impressive what's happening there!

**Comment by dsifford** [comment]
Woot-woot!! You da man! 😄 🎉 

Thanks! It's definitely coming along! 

I recently became aware of the [mobx state management library](https://mobxjs.github.io/mobx/) and oh...my....gosh using the observable pattern has changed my life! It cut the amount of code I had probably in half. Totally a gamechanger!

Thanks again for the quick fix. Hope you're doing well.

**Comment by dsifford** [comment]
@fbennett I'm still experiencing the issue even after the update. Any ideas?

**Comment by dsifford** [comment]
![image](https://cloud.githubusercontent.com/assets/5240018/17796457/93186f04-658e-11e6-9fde-4568e4a9c26a.png)

**Comment by dsifford** [comment]
note `processor_version` on line 3826 is `1.1.114`

**Comment by fbennett** [comment]
That's not good---but the demo works correctly with the new version, so maybe something is up with your processor input? Not sure what the problem could be---I tried a couple of things in the test framework, but no luck.

**Comment by dsifford** [comment]
Standby for a sec and I'll show you the exact shape of the data being given to `processCitationCluster`

**Comment by dsifford** [comment]
`processCitationCluster` takes 3 args. In my case, I those args `citation`, `before` and `after`.

With these assumptions:
- 1 citation already exists in the document with the HTML id `1kcmutkbg2`
- The new citation is being placed before it.

#### citation
```js
{
    citationItems: [
        { id: '1884heikc' },
    ],
    properties: {
         noteIndex: 0,
    }
}
```

### before
```js
[]
```

### after
```js
[
    ['1kcmutkbg2', 0],
]
```

**Comment by dsifford** [comment]
In return, I get this data:

```js
[
    {
        bibchange: false,
        citation_errors: []
    },
    [
        [0, '<sup>1</sup>',  '20lb7td6lm' ],
        [1, '<sup>2,3,1</sup>', '1kcmutkbg2' ],
]
```

**Comment by fbennett** [comment]
Thanks - there may be a tweak I can do to test input that will reproduce it.

**Comment by dsifford** [comment]
Thanks so much. Fingers crossed it's something easy 🙏

**Comment by fbennett** [comment]
It's certainly an issue in the code I just touched. Mid-commite now, will
write again later.

**Comment by dsifford** [comment]
@fbennett You are the bug-squishing master!

**Comment by fbennett** [comment]
Wasn't able to reproduce the fault, unfortunately. I had noteIndex set to non-zero values in my earlier tests. Changing it to 0 as in your input (which is correct for in-text references) didn't break things.

**Comment by dsifford** [comment]
Any idea what would cause the same input to produce a different output? Side effects?

**Comment by dsifford** [comment]
I have no problem investigating on my own, but I just need some sort of hint on where I should be looking. Lots of stuff going on in that particular function.

**Comment by dsifford** [comment]
Where does the resorting occur? `setSortKeys?`

I'm looking through `updateItems` now, hopefully I'm on the right track....

![image](https://cloud.githubusercontent.com/assets/5240018/17828093/4fdc14c0-6656-11e6-9f3f-54f967addceb.png)

**Comment by dsifford** [comment]
Ahhhhhhhhh! I figured it out!

It's because my `sys` getter was getting an observable map of the citation item, rather than a plain JS object. Somewhere in your code, it freaks out and can't handle a map representation of the data (I have no idea where). Not sure why, seeing as the observable map can be transacted the same way as an object (each key in the map is outfitted with `getters` for just that purpose).

Anyway, I guess it's fixed now. I just had to make a the function return a deep copy of the citation and transform it into an object prior to giving it to citeproc.

**Comment by fbennett** [comment]
Terrific! Glad to hear you've been able to pin it. I have errands this
afternoon, but I'll follow up later---the processor shouldn't stumble on
input that it can deal with.

**Comment by dsifford** [comment]
Correction on my last: It was not due to my `sys` getter, but rather was due to the following reason.

Because I need to keep my state updated with the processor state on a millisecond-to-millisecond basis, if at any time there is a change in the processor's `citationByIndex` array, I would take it and save it as an observable array.

The issue was that it wasn't creating a copy of the array, but rather saving it as a reference. Thus, when `citationByIndex` was made observable, it was also made observable within `CSL.Engine` instance.

The solution was to be sure to create a deep clone of `citationByIndex` (unfortunately, by doing `clone = JSON.parse(JSON.stringify(citationByIndex))` to completely dereference it from the processor. 

There's no way that you'll be able to guard for this unless you just use observables yourself (which I would recommend highly) but that would be a large project all on its own to convert to that pattern.

This issue is probably good to call resolved now. Thanks for the help!
